### PR TITLE
Add applications bolt implementation

### PR DESCRIPTION
**Pull request by jirevwe**

### CODE REVIEW

File: datastore/bolt/application.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+	return nil
+}
+
+func (a *appRepo) CountGroupApplications(tx context.Context, groupID string) (int64, error) {
+	return 0, nil
+}
+
+func (a *appRepo) createUpdateApplication(app *convoy.Application) error {
+	return a.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(a.bucketName))
```
Comment (danvixent) [inline datastore/bolt/application.go:101]: I figured we can still use constant for this thing

```go
const eventDeliveryBucketName = "eventdeliveries"
const eventBucketName = "events"

```

File: datastore/bolt/group.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	var group *convoy.Group
+	err := g.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket([]byte(g.bucketName))
+
+		grp := b.Get([]byte(gid))
+		if grp == nil {
+			return fmt.Errorf("group with id (%s) does not exist", gid)
+		}
+
+		var _grp *convoy.Group
```
Comment (danvixent) [inline datastore/bolt/group.go:78]: you ca reference the group you declared previously.

```go
err := json.Unmarshal(buf, group)
if err != nil {
   return err
}

```

File: datastore/bolt/application.go (hunk 1 above)
Comment (jirevwe) [reply datastore/bolt/application.go:273]: Nah, just add it to the struct since we'd also need it to create the bucket

Comment (danvixent) [reply datastore/bolt/application.go:273]: e no go work 😭 , apparently you can't have multiple root buckets, but you can have multiple sub buckets in the root buckets. check my pr for reference.

File: cmd/main.go [hunk 3]
```diff
@@ -87,6 +88,11 @@ func main() {
 				return err
 			}
 
+			bolt, err := bolt.New(cfg)
```
Comment (Subomi) [inline cmd/main.go:91]: @jirevwe we should define `datastore.NewDB(cfg)`. It will do the if/else and return `datastore.DatabaseClient`

**Review by Subomi — DISMISSED** [review]

File: cmd/main.go (hunk 3 above)
Comment (danvixent) [reply cmd/main.go:91]: i see a import cycle here, bolt -> datastore.Application, datastore -> bolt.New

Comment (jirevwe) [reply cmd/main.go:91]: Hmm, let me check

File: datastore/bolt/application.go [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+		if pErr != nil {
+			return pErr
+		}
+
+		return nil
+	})
+}
+
+func (a *appRepo) DeleteGroupApps(ctx context.Context, gid string) error {
+	return a.db.View(func(tx *bbolt.Tx) error {
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply datastore/bolt/application.go:237]: Good catch.

File: cmd/main.go (hunk 3 above)
Comment (jirevwe) [reply cmd/main.go:91]: Yep, there's an import cycle. What do you think @Subomi?

**Review by Dotunj — APPROVED** [review]

File: datastore/bolt/group.go [hunk 5]
```diff
@@ -1,114 +1,134 @@
...
-		mErr := json.Unmarshal(grp, &_grp)
-		if mErr != nil {
-			return mErr
+		var temp *datastore.Group
+		err := json.Unmarshal(grp, &temp)
+		if err != nil {
+			return err
 		}
-		group = _grp
+		group = temp
```
Comment (danvixent) [inline datastore/bolt/group.go:82]: you can just reference the `var group *datastore` here 
```go
err := json.Unmarshal(grp, &group)
if err != nil {
    return err
}
```

File: datastore/bolt/group.go [hunk 6]
```diff
@@ -1,114 +1,134 @@
...
 	err := g.db.View(func(tx *bbolt.Tx) error {
-		id := name + ":" + gid
+		b := tx.Bucket([]byte(g.bucketName))
 
-		grp := tx.Bucket([]byte(bucketName)).Get([]byte(id))
+		grp := b.Get([]byte(gid))
 		if grp == nil {
-			return fmt.Errorf("group with id (%s) does not exist", gid)
+			return datastore.ErrGroupNotFound
 		}
```
Comment (danvixent) [inline datastore/bolt/group.go:75]: we can call this `buf` instead as that name denotes the byte slice property better

File: datastore/bolt/application.go [hunk 7]
```diff
@@ -3,70 +3,319 @@ package bolt
...
+			data.TotalPage = int64(math.Ceil(float64(b.Stats().KeyN) / float64(perPage)))
+			data.Total = int64(b.Stats().KeyN)
+		} else {
+			total, err := a.CountGroupApplications(ctx, gid)
+			if err != nil {
+				return nil
+			}
+
+			data.TotalPage = int64(math.Ceil(float64(total) / float64(perPage)))
+			data.Total = int64(total)
```
[the remark this answers is not in the document]
Comment (jirevwe) [reply datastore/bolt/application.go:106]: This is not correct though. I added a test case to validate it and it's not correct

**Review by danvixent — APPROVED** [review]

File: datastore/bolt/application.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (danvixent) [reply datastore/bolt/application.go:108]: nicee.

**Pull request merged**